### PR TITLE
grpc: Use cmake for build and bump to 1.31.1

### PR DESCRIPTION
**Pull request by kabel**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Reboots #59195 with bump to 1.31.1 and build system fixes to use cmake.

### CODE REVIEW

File: Formula/grpc.rb [hunk 1]
```diff
@@ -21,25 +21,44 @@ class Grpc < Formula
...
+    mkdir "cmake/build" do
+      args = %w[
+        ../..
+        -DBUILD_SHARED_LIBS=ON
+        -DgRPC_BUILD_TESTS=OFF
+        -DgRPC_INSTALL=ON
+        -DgRPC_ABSL_PROVIDER=package
+        -DgRPC_CARES_PROVIDER=package
+        -DgRPC_PROTOBUF_PROVIDER=package
+        -DgRPC_SSL_PROVIDER=package
```
Comment (SMillerDev) [inline Formula/grpc.rb:42]: Are these upstream's way of defining system dependencies?

Comment (kabel) [reply Formula/grpc.rb:42]: Yes. See https://github.com/grpc/grpc/blob/master/BUILDING.md#dependency-management

Comment (SMillerDev) [reply Formula/grpc.rb:42]: Just checking since package could be explained either way without context.

**Review by SMillerDev — APPROVED** [review]
Thanks @kabel  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Pull request closed without merging**